xapian-applications/omega/handler_libarchive.cc

```diff
@@ -0,0 +1,122 @@
+/** @file handler_libarchive.cc
+ * @brief Extract text and metadata using libarchive.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "metaxmlparse.h"
+#include "opendocparse.h"
+#include "str.h"
+
+#include<archive.h>
+#include<archive_entry.h>
+
+#include <cstring>
+
+using namespace std;
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	struct archive* archive_obj;
+	struct archive_entry* entry;
+	int status_code;
+	const char* file = filename.c_str();
+	archive_obj = archive_read_new();
+	archive_read_support_filter_all(archive_obj);
+	archive_read_support_format_all(archive_obj);
+	status_code = archive_read_open_filename(archive_obj, file, 10240);
+
+	if (status_code != ARCHIVE_OK) {
+	    error = "Libarchive failed to open the file " + filename;
+		return false;
+	}
+
+	size_t total;
+	ssize_t size;
+	string s;
+
+	// extracting data from content.xml and styles.xml
+	while (archive_read_next_header(archive_obj, &entry) == ARCHIVE_OK) {
+	    if (strcmp(archive_entry_pathname(entry), "content.xml") == 0) {
+		total = archive_entry_size(entry);
+		char buf1[total];
+		size = archive_read_data(archive_obj, buf1, total);
+
+		if (size <= 0) {
+		    error = "Libarchive was not able to extract data from "
+		    "content.xml";
+		    return false;
+		}
+
+		s += str(buf1);
```

Comment:
It would be better using different strings to store the information extracted as we cannot assume any particular order in which libarchive browses on the file.

Reply:
Also you really need to use the length read (which is in `size`), not rely on there being a terminating zero byte (which the libarchive documentation doesn't mention being added).  You'll often happen to have a zero byte after the read data as memory is typically zero-initialised on first allocation in modern machines, but in a longer running process as heap memory gets reused this will likely end up with extra junk getting appended to the read archive member data.

Reply:
I have used string resizing as Olly suggested so char array is not being used anymore.

Also, I will use two different strings for `content.xml` and 'styles.xml`, I haven't done this yet, will make the change and push it again.